File: Medium/Top K Frequent Elements/topkfrequentelements.java
import java.util.*;

class IntegerPair implements Comparable<IntegerPair> {// stores coordinates
    private Integer _first, _second;
  
    public IntegerPair(Integer f, Integer s) {
      _first = f;
      _second = s;
    }
  
    public int compareTo(IntegerPair o) {
      if (!this.first().equals(o.first())) return this.first() - o.first();
      else return this.second() - o.second();
    }
  
    Integer first() {return _first;}
  
    Integer second() {return _second;}
}

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer,Integer> frequencies = new HashMap<>();
        for (int n: nums){
            if (!frequencies.containsKey(n)) frequencies.put(n,1);
            else frequencies.put(n,frequencies.get(n)+1);
        }
        ArrayList<IntegerPair> lst = new ArrayList<>();
        // first coordinate is frequency, second coordinate is number
        for (int key: frequencies.keySet()) lst.add(new IntegerPair(frequencies.get(key), key)); 
        Collections.sort(lst, Collections.reverseOrder());
        int[] ans =  new int[k];
        for (int i = 0; i < k; i++) ans[i] = lst.get(i).second();
        return ans;
    }
}

File: Medium/Top K Frequent Elements/topkfrequentelements.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;
// sort by frequency in descending order and by number in descending order
// customComparator function must be placed out of Solution class
bool customComparator(const pair<int,int>&a, const pair<int,int>&b){
    if (a.second != b.second) return a.second > b.second;
    return a.first > b.first;
}

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int,int> frequencies;
        for (int n: nums){
            if (frequencies.count(n) == 1) frequencies[n]++;
            else frequencies[n] = 1;
        }
        vector<pair<int,int>> lst;
        for (pair<int,int> p: frequencies) lst.push_back(p);
        sort(lst.begin(),lst.end(),customComparator);
        vector<int> ans;
        for (int i = 0; i < k; i++) ans.push_back(lst[i].first);
        return ans;
    }
};